File: marshaller.py
import struct
from messages.marshalable import Marshalable
from messages.balance_msg import *
from messages.dw_msg import *
from messages.create_new_account_output import * 
from messages.create_new_account_input import *
from enum import Enum
from bank_account import BankAccount
from unmarshaller import *

class Marshaller:
	
	@staticmethod
	def marshal_string(data: str):
		"""
		Marshal string using the string encode method.
		Keyword arguments:
		data: string to be marshalled

		Returns: 
		Marshalled string output/type error if data is not a string object.
		"""

		if not isinstance(data, str):
			print("Error! marshal_string called on non-string value!")
			raise TypeError

		result = bytes()

		str_len = len(data)
		str_len = Marshaller.marshal_int(str_len, 4)
		content = data.encode('utf-8')

		result += str_len + content

		return result


	@staticmethod
	def marshal_int(data: int, num_bytes: int = 8):
		"""
		Marshal int using the struct pack method.
		Keyword arguments:
		data: value to be marshalled
		num_bytes: number of bytes in output (4 or 8)

		Returns: 
		Marshalled int output/type error if data is not an int/float object.
		"""
		
		if isinstance(data, float):
			print("marshal_int called on float value! Converting to int...")
			data = int(data)
		
		if not isinstance(data, int):
			print("Error! marshal_int called on non-int value!")
			raise TypeError

		if num_bytes == 4:
			return struct.pack('<i', data)

		return struct.pack('<Q', data)

	
	@staticmethod
	def marshal_float(data):
		"""
		Marshal float using the struct pack method.
		Keyword arguments:
		data: value to be marshalled

		Returns: 
		Marshalled float output/type error if data is not an int/float object.
		"""

		if isinstance(data, int):
			print("marshal_float on int value! Converting to float...")
			data = float(data)

		if not isinstance(data, float):
			print("Error! marshal_float called on non-float value!")
			raise TypeError
			
		return struct.pack('<d', data)
	

	@staticmethod
	def marshal_enum(data):
		""""
		Marshal enum type object.
		Keyword arguments:
		data: value to be marshalled

		Returns: 
		Marshalled enum output/type error if data is not an enum object.
		"""

		if not isinstance(data, Enum):
			print("Error! marshal_enum called on non-enum value!")
			raise TypeError
		
		result = bytes()

		obj_type = type(data).object_type()
		result += Marshaller.marshal_int(obj_type, 4)
		result += Marshaller.marshal_int(data.value, 4)

		return result
	

	@staticmethod
	def marshal_object(data):
		"""
		Marshal objects derived from Marshalable class.
		Keyword arguments:
		data: object to be marshalled

		Returns: 
		Marshalled object output/type error if object doesn't have Marshalable as parent class.
		"""

		data_class = type(data)

		if not issubclass(data_class, Marshalable):
			print("Error! marshal_object called on non-marshalable object!")
			raise TypeError

		result = bytes()
		
		obj_type = data_class.object_type()
		result += Marshaller.marshal_int(obj_type, 4)

		fields_dict = data.get_fields()
		fields_type_dict = data_class.get_field_types()

		for field_id, field_val in fields_dict.items():
			field_type = fields_type_dict[field_id]
			result += Marshaller.marshal_int(field_id, 4)

			if field_type == int:
				result += Marshaller.marshal_int(field_val)
			elif field_type == float:
				result += Marshaller.marshal_float(field_val)
			elif field_type == str:
				result += Marshaller.marshal_string(field_val)
			elif field_type == Enum:
				result += Marshaller.marshal_enum(field_val)

		return result


def compile_message(message_id: int, object: Marshalable) -> bytes:
	"""
	Compile final message to be sent over.
	Keyword arguments:
	message_id: id from the response object being sent to client
	object: response object to be marshalled

	Returns: 
	Marshalled object output/type error if object doesn't have Marshalable as parent class.
	"""

	if not issubclass(type(object), Marshalable):
		print("Object sent to marshaller is inherited from Marshalable!")
		raise TypeError

	result = bytes()
	
	marshalled_message_id = Marshaller.marshal_int(message_id, 4)
	marshalled_object = Marshaller.marshal_object(object)

	result += marshalled_message_id + marshalled_object

	message_len = len(result)
	marshalled_message_len = Marshaller.marshal_int(message_len, 4)
	
	result = marshalled_message_len + result

	return result

File: client/marshaller.go
package main

import (
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"math"
)

// Interface to describe the functions to be implemented by the marshaller
type marshal_functions interface {
	marshal() []byte
	marshal_string(string_data string) []byte
	marshal_uint32(int_data uint32) []byte
	marshal_uint64(uint64_data uint64) []byte
	marshal_float(float_data float64) []byte
	marshal_struct(struct_data Marshalable) []byte
	marshal_enum(enum_data Marshalable) []byte
}

// Dummy struct to represent a marshaller object
type marshaller struct {
}

// Unimplemented function included in case of required extension at a later date
func (m marshaller) marshal() []byte {
	bs := make([]byte, 8)
	return bs
}

/*
Marsal unsigned 32-bit integer using binary package
Keyword arguments:
int_data: 32-bit value to be marshalled

Returns:
Marshalled 4-byte output
*/
func (m marshaller) marshal_uint32(int_data uint32) []byte {
	data := make([]byte, 4)
	binary.LittleEndian.PutUint32(data, int_data)
	return data
}

/*
Marsal unsigned 64-bit integer using binary package
Keyword arguments:
uint64_data: 64-bit value to be marshalled

Returns:
Marshalled 8-byte output
*/
func (m marshaller) marshal_uint64(uint64_data uint64) []byte {
	data := make([]byte, 8)
	binary.LittleEndian.PutUint64(data, uint64_data)
	return data
}

/*
Marsal 64-bit float using binary package
Keyword arguments:
float_data: 64-bit float to be marshalled

Returns:
Marshalled 8-byte output
*/
func (m marshaller) marshal_float(float_data float64) []byte {
	data := make([]byte, 8)
	var uint64_float uint64 = math.Float64bits(float_data)
	binary.LittleEndian.PutUint64(data, uint64_float)
	return data
}

/*
Marsal string using binary package
Keyword arguments:
string_data: string to be marshalled

Returns:
Marshalled string output
*/
func (m marshaller) marshal_string(string_data string) []byte {
	var str_len uint32 = uint32(len(string_data))
	var len_marshalled []byte = m.marshal_uint32(str_len)
	var str_marshalled []byte = []byte(string_data)
	return append(len_marshalled, str_marshalled...)
}

/*
Marsal objects implementing the Marshalable interface using binary package
Keyword arguments:
struct_data: struct to be marshalled

Returns:
Marshalled struct output
*/
func (m marshaller) marshal_struct(struct_data Marshalable) []byte {
	data := make([]byte, 0)
	var obj_type int = struct_data.object_type()
	data = append(data, m.marshal_uint32(uint32(obj_type))...)
	var fields_map map[int]any = struct_data.get_fields()
	fmt.Println(fields_map)
	for field_id, field_val := range fields_map {
		data = append(data, m.marshal_uint32((uint32(field_id)))...)
		switch v := field_val.(type) {
		case uint64:
			data = append(data, m.marshal_uint64(field_val.(uint64))...)
		case CurrencyType: 
			data = append(data, m.marshal_uint32(uint32(field_val.(CurrencyType).object_type()))...)
			data = append(data, m.marshal_uint32(field_val.(CurrencyType).getValue())...)
		case float64:
			data = append(data, m.marshal_float(field_val.(float64))...)
		case string:
			data = append(data, m.marshal_string(field_val.(string))...)
		default:
			fmt.Println("An error occurred %v "+v.(string), field_val)
		}
	}
	return data
}

/*
Marsal enum using binary package
Keyword arguments:
enum_data: enum to be marshalled

Returns:
Marshalled 8-byte output
*/
func (m marshaller) marshal_enum(enum_data Marshalable) []byte {
	data := make([]byte, 0)
	data = append(data, m.marshal_uint32(uint32(enum_data.object_type()))...)
	data = append(data, m.marshal_uint32(enum_data.get_fields()[0].(uint32))...)
	return data
}

/*
Compile final message to be sent over the network
Keyword arguments:
m: marshaller that implements the marshal_functions interface
message_id: id set by client
object: response object (implementing Marshalable) to be marshalled

Returns:
Complete marshalled message
*/
func compile_message(m marshal_functions, message_id int, object Marshalable) []byte {
	result := make([]byte, 0)
	result = append(result, m.marshal_uint32(uint32(message_id))...)
	result = append(result, m.marshal_struct(object)...)
	var message_len uint32 = uint32(len(result))
	result = append(m.marshal_uint32(message_len), result...)

	return result
}

// Testing function
func printDetails(m marshal_functions) {
	fmt.Println((m.marshal_string("Hello There!")))
	fmt.Println((m.marshal_uint64(1024)))
	fmt.Println((m.marshal_float(5.1)))
	var qMessage Marshalable = QueryBalanceMessage{"Sid", uint64(10853693087894514759), "password123"}
	var dMessage Marshalable = DepositMessage{DWBaseMessage{"Sid", uint64(10853693087894514759), "password123", SGD, 105.5}}
	var cMessage Marshalable = CreateBankAccountMessage{"Sid", "password", 105.5, USD}
	fmt.Println((m.marshal_struct(qMessage)))
	fmt.Println((hex.EncodeToString(m.marshal_struct(dMessage))))
	fmt.Println(compile_message(m, 20, cMessage))
	fmt.Println((hex.EncodeToString(compile_message(m, 12, qMessage))))
	fmt.Println((compile_message(m, 16, dMessage)))
}

// Function to generate the registry of all the custom structs used in code
func generateRegistry(r *Registry) error {
	//r.Put(&foo{})
	r.Put(&CloseAccountMessage{})
	r.Put(&CloseAccountResponse{})
	r.Put(&QueryBalanceMessage{})
	r.Put(&QueryBalanceResponse{})
	r.Put(&CreateBankAccountMessage{})
	r.Put(&CreateBankAccountResponse{})
	r.Put(&DepositMessage{})
	r.Put(&WithdrawMessage{})
	r.Put(&ErrorMessage{})
	r.Put(&MonitorResponse{})
	r.Put(&RegisterMonitorMessage{})
	r.Put(&RegisterMonitorOutput{})
	r.Put(&TransferMessage{})
	r.Put(CurrencyType(0))
	return nil
}

// func main () {
// 	m := marshaller{}
// 	registry, err := NewRegistry(generateRegistry)
// 	if err != nil {
// 		fmt.Printf("error: %s\n", err.Error())
// 		return
// 	}
// 	fmt.Println(reflect.TypeOf("a"))
// 	um := unmarshaller{registry}
// 	// value, err := registry.Get(602)
// 	// if err != nil {
// 	// 	fmt.Printf("error: %s\n", err.Error())
// 	// 	return
// 	// }
// 	// i := value.Unwrap()
// 	// var object Marshalable = getMarshalableObject(i)
// 	// fmt.Println(object.object_type())
// 	printDetails(m)
// 	printUnmarshalDetails(um)
// }
